dss/util/iterators.py

```diff
@@ -0,0 +1,177 @@
+from itertools import tee
+from typing import Iterator, TypeVar, Optional, Tuple, Iterable, NamedTuple
+
+T = TypeVar('T')
+OT = Optional[T]
+Column = Iterator[T]
+
+
+class zipalign(Iterator['zipalign.Row']):
```

Comment:
Unfortunately, the alternative does not satisfy the resumability requirement. I'm afraid it would become more complicated than the original solution if we had to add that.

We need to be able to reinitialize the `zipalign` iterator with external state because the visitation worker can be interrupted between rows. The original solution only needs to persist the Row object itself.  The alternative proposal would also have to persist the hidden state introduced by `peekable`, i.e. the buffer that stores the element returned by the wrapped iterator's next().

Reply:
Can you help me understand this resumability problem? Let's hop on a call once you're back and you can show me.

Again, my concern here is clarity of the code, not to compete on alternatives. The 'example' code above just demonstrates that you can fulfill the same logical requirements concisely (~15-20 LOC). There are at least two other ways of doing it that would be about the same in terms of readability. You don't need to use `peekable`.

Reply:
+1 on the call. However, to give you a head-start: like all lambdas, the visitation lambda is limited to 5min. In those 5 min, you only make it so far through the iterators that yield the bundle keys in each replica. So you have to snapshot some state to resume from when the worker lambda gets invoked again by the visitation StepFunction machinery. In this case you need to snapshot that state for each row or you risk emitting log message more than once.

Reply:
Log messages about inconsistencies that is.

Reply:
I agree, 
> "I like to promote generalized functionality into its own prominent location even though there is only one usage of it. Separating it out sends a strong signal to other developers about the generality of the component and the loose coupling. Keeping it close to the usage makes it easier to accidentally increase the coupling."